### ISSUE TITLE
How to use getAll() in my own script?

### ISSUE THREAD

**Issue by wsw70** [11ty/eleventy#871]
I would like to process the content of my blog posts as part of a JS script (specifically I will use Vue.js to dynamically enable some tags and update the content accordingly).

To do so, I need in my `script.js` called from `index.html` to have access to the content of all processed posts. I think that the [`getAll()`](https://www.11ty.dev/docs/collections/#getall()) function was intended just for that.

How in practical terms can I use it in my script?  Is it possible to
- somehow call `myContent = getAll() ` in `script.js` ?
- or have it called in `.eleventy.js` and access the resulting Array in `script.js`?

**Comment by rphunt** [comment]
"To do so, I need in my script.js called from index.html to have access to the content of all processed posts. I think that the getAll() function was intended just for that."
I'm not a Vue person, but maybe I can dig into one possible issue.

When you say "script.js called from index.html" do you mean when index.html is loaded in a browser? 

If so, I'm not sure where getAll() or .eleventy.js come into the picture. Those are only used when Eleventy renders the static pages. Neither is available in real time in a live site.

Maybe what would work is making a template that generates a file in a data format that Vue can access in real time. It's not something I've tried myself, tho.

Basically, Eleventy can <em>use</em> API's but it can't <em>be</em> one (yes, I realize you're not talking about an <em>actual</em> API, but I think the comparison is valid).

But maybe I don't understand, and somehow you're envisioning involving Vue in the rendering process. Hopefully someone else has a more educated perspective on that.

**Comment by wsw70** [comment]
@rphunt thanks for the answer and sorry for not having been more clear

> If so, I'm not sure where getAll() or .eleventy.js come into the picture. Those are only used when Eleventy renders the static pages. Neither is available in real time in a live site.

My first idea (the first bullet point) was to use eleventy after the rendering. I thought that it would leave behind an API (a JS file) which can be called. This is not the case as I now understand that eleventy creates the static version of the site and that's perfectly fine.

> Maybe what would work is making a template that generates a file in a data format that Vue can access in real time.

This was the second idea → to create  during the processing of the source files a left-behind file which would contain all information about the posts. I am thinking about a JSON-formatted file which I could then `fetch` in my static page code (from Vue) and process it from there.

To do so I would need:
- a hook into the generation process to plug in "now get all the data for all pages (collections, tags, content, ...)"
- a function which does the collection (called in the hook above). I think that `getAll()` could be that function
- a way to write to `_site` the JSON file with that data

I hope that it is more clear now.

**Comment by rphunt** [comment]
Well, you're familiar with <b>getAll()</b> which implies you're familiar with <b>addCollection</b>. To me, that's half the battle. Of course, getAll() contains a LOT of data, I don't think you want a full dump of it. If you ever console.log it, you'll see. There's probably a lot of stuff you would not need (and a lot is redundant).

What I would probably do is create collections of the data I want to make available to Vue, then make a template to render them out in a JSON format, either as discrete files or one data set. Note that access to global data is not direct, but it <em>is</em> in there.

[https://github.com/11ty/eleventy/issues/846]

Of course,  if you're using nodeJS, all Node's functionality is available. So you might be able to have .eleventy.js actually write a JSON file directly. Again, this is not something I have tried, but now I'm curious.

What I would also do is probably post a new question here to see if someone has already addressed spitting out JSON from an Eleventy collection. There might already be a solution out there.

**Comment by rphunt** [comment]
I just did a VERY quick and dirty test. I have a global data file dogs.json:
<pre>
[
    "collie",
    "pug",
    "poodle",
    "beagle",
    "retriever",
    "shepherd",
    "spaniel",
    "terrier"
]
</pre>

Pulled it out from getAll(), and was able to create a new json file from that:

<pre>
eleventyConfig.addCollection('dogsCollection', function (collection) {
    var dogs = collection.getAll()[0].data.dogs;
    fs.writeFileSync('./_site/test.json', JSON.stringify(dogs));
    return dogs;
});
</pre>

Don't know how much this helps.

**Comment by wsw70** [comment]
> Don't know how much this helps.

Thanks a lot. I actually got to the same point by poking around `.addCollection()` and this will be the way to go. It is exactly what I was looking for - thanks again for the discussion which cleared up a lot.

I see a problem in the serialization of the whole set of data, opened a bug for that (https://github.com/11ty/eleventy/issues/872)

**Comment by Ryuno-Ki** [comment]
In case someone comes across this later:
@cfjedimaster blogged about how to [Adding Search to your Eleventy Static Site with Lunr](https://www.raymondcamden.com/2019/10/20/adding-search-to-your-eleventy-static-site-with-lunr) which involved generating a JSON with all the content (I'm about to try the Lunr approach myself later this year).
